internal/keepers/ocr.go

```diff
@@ -100,38 +100,67 @@ func (k *keepers) Report(ctx context.Context, rt types.ReportTimestamp, _ types.
-	// select, verify, and build report
-	toPerform := make([]ktypes.UpkeepResult, 0, 1)
-	for _, key := range keys {
-		upkeeps, err := k.service.CheckUpkeep(ctx, key)
-		if err != nil {
-			return false, nil, fmt.Errorf("%w: failed to check upkeep from attributed observation: %s", err, lCtx)
-		}
+	// No keys found for the given keys
+	if len(keys) == 0 {
+		k.logger.Printf("OCR report completed successfully with no eligible keys: %s", lCtx)
+		return false, nil, nil
+	}
 
-		// No upkeeps found for the given key
-		if len(upkeeps) == 0 {
-			continue
-		}
+	// Check all upkeeps from the given observation
+	checkedUpkeeps, err := k.service.CheckUpkeep(ctx, keys...)
+	if err != nil {
+		return false, nil, fmt.Errorf("%w: failed to check upkeeps from attributed observation: %s", err, lCtx)
+	}
 
-		if len(upkeeps) > 1 {
-			return false, nil, fmt.Errorf("unexpected number of upkeeps returned for %s key, expected 1 but given %d", key, len(upkeeps))
-		}
+	// No upkeeps found for the given keys
+	if len(checkedUpkeeps) == 0 {
+		k.logger.Printf("OCR report completed successfully with no successfully checked upkeeps: %s", lCtx)
+		return false, nil, nil
+	}
 
-		if upkeeps[0].State == ktypes.Eligible {
-			// only build a report from a single upkeep for now
-			k.logger.Printf("reporting %s to be performed: %s", upkeeps[0].Key, lCtx.Short())
-			toPerform = append(toPerform, upkeeps[0])
-			break
+	if len(checkedUpkeeps) > len(keys) {
+		return false, nil, fmt.Errorf("unexpected number of upkeeps returned for %s key, expected max %d but given %d", key, len(keys), len(checkedUpkeeps))
+	}
+
+	// Collect eligible upkeeps
+	keyToUpkeep := make(map[string]ktypes.UpkeepResult)
+	eligibleUpkeeps := make([]ktypes.UpkeepKey, 0, len(checkedUpkeeps))
+	for _, checkedUpkeep := range checkedUpkeeps {
+		if checkedUpkeep.State == ktypes.Eligible {
+			eligibleUpkeeps = append(eligibleUpkeeps, checkedUpkeep.Key)
+			keyToUpkeep[string(checkedUpkeep.Key)] = checkedUpkeep
 		}
 	}
 
 	// if nothing to report, return false with no error
-	if len(toPerform) == 0 {
-		k.logger.Printf("OCR report completed successfully with no upkeeps added to the report: %s", lCtx)
+	if len(eligibleUpkeeps) == 0 {
+		k.logger.Printf("OCR report completed successfully with no eligible upkeeps: %s", lCtx)
 		return false, nil, nil
 	}
 
+	// Get upkeep gas limit
+	upkeepsInfo, err := k.service.GetUpkeep(ctx, eligibleUpkeeps...)
+	if err != nil {
+		return false, nil, fmt.Errorf("%w: failed to get upkeeps info: %s", err, lCtx)
+	}
+
+	// Prepare upkeeps to build report
+	var reportCapacity uint32
+	toPerform := make([]ktypes.UpkeepResult, 0, len(upkeepsInfo))
+	for i, upkeepInfo := range upkeepsInfo {
+		upkeepResult, ok := keyToUpkeep[string(eligibleUpkeeps[i])]
+		if !ok {
+			return false, nil, fmt.Errorf("upkeep with key %s not found in the info list", string(eligibleUpkeeps[i]))
+		}
+
+		if reportCapacity+upkeepInfo.ExecuteGas > k.reportGasLimit {
+			break
+		}
```

Comment:
we'll need to account for gas overhead also somewhere. Currently if upkeep gas limit is 5M, then we send txs with 5.3M. 
300K overhead per upkeep is a good margin to add (can probably be tightened as incremental upkeeps don't need full 300K overhead, but should be good to start with)

Reply:
The gas overhead should be configured on the chainlink node side where transmitter is implemented

Reply:
yes, but currently since transmitter does not take dynamic gas limits, it always uses a fixed one here we will configure it on the chainlink node side as an off-chain config

Reply:
Ok, so what do you suggest? We have nothing to do with the gas overhead here IMO. This logic is needed to calculate the report capacity.

Reply:
So we need to account for the gas overheads somewhere, can discuss the best place to put that.

From contract side, we need to set the gas limit to be sum of upkeeps gas limits + overhead. Overhead required scales with number of upkeeps, 300K per upkeep is an upper bound

Reply:
Ahhhh. I get it. The overhead depends on the number of upkeeps. Not just the general overhead. That could be broken down to two different 'overheads': the per upkeep overhead and the general transmit overhead. Though it may not be worth our efforts to split them like this.